### PR TITLE
Tuples/Lists can now be inputs/outputs to script and other simple fixes.

### PR DESCRIPTION
**Pull request by zdevito**
* Fix the necessary pathways so that tuples and lists can be inputs to the script.

* prevent linear algebra functions from being run in shape prop because
they frequently will error out for nonsense data.

* favor schema-driven python input conversion where possible.
remaining cases where we directly create Stacks without schema are
only for debugging

* Make the error messages when calling script/trace functions more pythonic

* Simplify FlattenTuples -- now that tuples are supported we can choose to only flatten tuples when needed. This may have to be revisited pending onnx test results, but is necessary for making tuple io work.

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -2021,6 +2021,28 @@ def checkScript(self, script, inputs, optimize=True, outputs=None, name='func',
 
         return ge
 
+    def test_tuple_io(self):
+        def stuff(x):
+            # type: (Tuple[Tensor, Tensor]) -> Tuple[Tensor, Tensor]
+            a, b = x
+            return b, a
+
+        a = (torch.ones(3), torch.ones(3))
```
Comment (apaszke) [inline test/test_jit.py]: It would be better to use distinct values for this test, because a function that returns `x` unchanged would pass it too.

File: test/test_jit.py [hunk 2]
```diff
@@ -6657,13 +6673,9 @@ def test_passing_unknown_kwargs(self):
...
-        with self.assertRaisesRegex(
-            RuntimeError,
-            "Lists and strings are not supported yet"
-        ):
-            a, b = torch.ones(5), torch.zeros(5)
-            output = torch.ops.aten.stack([a, b])
-            self.assertEqual(output, torch.ones(10))
+        a, b = torch.ones(5), torch.ones(5)
+        output = torch.ops.aten.cat([a, b])
+        self.assertEqual(output, torch.ones(10))
```
Comment (apaszke) [inline test/test_jit.py]: Again, would be better to have a test that actually cares that it was `cat([a, b])` and not `cat([b, a])`.

File: torch/csrc/jit/passes/lower_tuples.cpp [hunk 3]
```diff
@@ -5,119 +5,35 @@
...
-  for (Node* n : block->nodes()) {
-    for (Block* b : n->blocks()) {
-      EnsureNoTuples(b);
-    }
-    for (Value * o : n->outputs()) {
-      JIT_ASSERTM(o->type()->kind() != TypeKind::TupleType,
-                  "Couldn't lower all tuples. This is an error because "
-                  "they're not implemented in the interpreter just yet.");
-    }
+  for(auto n : block->nodes()) {
```
Comment (apaszke) [inline torch/csrc/jit/passes/lower_tuples.cpp]: Why do we no longer flatten tuples that are inputs or outputs to blocks? This is actually important, because we use it to fix up graphs that take/return tuples into flat arg lists for ONNX.

File: torch/csrc/jit/passes/shape_analysis.cpp [hunk 4]
```diff
@@ -191,7 +191,15 @@ bool isValidReturnForRunning(Value* v) {
       v->type()->isSubtypeOf(NumberType::get());
 }
 
+std::unordered_set<Symbol> cannot_propagate_shape_by_running_it = {
+  aten::gesv,
+  aten::inverse,
+};
+
 bool canPropagateShapeByRunningIt(Node* node) {
+  if(cannot_propagate_shape_by_running_it.count(node->kind()))
```
Comment (apaszke) [inline torch/csrc/jit/passes/shape_analysis.cpp]: nit: `OperatorSet`? 😊

File: torch/csrc/jit/pybind_utils.h [hunk 5]
```diff
@@ -67,6 +67,16 @@ inline Stack toStack(const py::tuple& inputs) {
   return toIValue(inputs).toTuple()->elements();
 }
 
+inline IValue toIValue(py::handle obj, const TypePtr& type);
+
+inline IValue createGenericList(py::handle obj, const TypePtr& elem_type) {
```
Comment (apaszke) [inline torch/csrc/jit/pybind_utils.h]: Why is this called a "generic list" and not a tuple, like everywhere else? We don't really have a concept of an IValue list.

File: torch/csrc/jit/script/module.h [hunk 6]
```diff
@@ -149,12 +149,9 @@ struct Method {
...
   }
 
-  const FunctionSchema& getSchema() const {
-    AT_ASSERT(schema != nullptr);
-    return *schema;
-  }
+  const FunctionSchema& getSchema() const;
 
-  std::string prettyPrintSchema() const {
+  std::string pretty_print_schema() const {
```
Comment (apaszke) [inline torch/csrc/jit/script/module.h]: Why the change in naming convention? We usually use camel case in the JIT, right?

**Review by apaszke — COMMENTED** [review]
LGTM, but there are a few suspicious things.

File: torch/csrc/jit/passes/lower_tuples.cpp (hunk 3 above)
Comment (zdevito) [reply torch/csrc/jit/passes/lower_tuples.cpp]: It is not always a good decision to flatten tuples (imagine a huge one, or a 2 element one where one is small and the other huge). We previously had to. I am going to change this so that we leave the old pass to completely flatten for onnx export (though disturbingly the onnx tests do not fail with the code as it), while we only typically run the new very simple pass.

File: torch/csrc/jit/pybind_utils.h (hunk 5 above)
Comment (zdevito) [reply torch/csrc/jit/pybind_utils.h]: Michael and I were discussing this: we have to have a fallback boxed list representation for lists of things that are not ints, float, or Tensors. This is just making this code robust to that future. Once Michaels mutable list patch lands, there will be a distinction between the Tuple in ivalue (const list), and the mutable boxed list type.

File: torch/csrc/jit/script/module.h (hunk 6 above)
Comment (zdevito) [reply torch/csrc/jit/script/module.h]: module.h is bound to python. Its other methods use python convention, so I am trying to incrementally make it more consistent. There are other changes but they were more invasive so I didn't do it yet.

File: test/test_jit.py [hunk 7]
```diff
@@ -6730,13 +6746,10 @@ def test_passing_unknown_kwargs(self):
 
     def test_passing_and_returning_lists(self):
         # Replace with actual test once we support lists.
```
Comment (goldsborough) [inline test/test_jit.py]: You can remove this comment now

File: torch/csrc/jit/passes/lower_tuples.cpp [hunk 8]
```diff
@@ -101,23 +101,55 @@ static void LowerTuples(Block* block) {
...
 }
 
-void LowerTuples(std::shared_ptr<Graph>& graph) {
-  LowerTuples(graph->block());
+void LowerAllTuples(std::shared_ptr<Graph>& graph) {
+  // we can't lower all tuples if are tuples in the input or outputs
+  // we need to check because LowerTuples will try to flatten them,
+  // editing the inputs/outputs to the graph in a way that is not desired
+  EnsureNoTuples(graph->inputs());
+  EnsureNoTuples(graph->outputs());
```
Comment (apaszke) [inline torch/csrc/jit/passes/lower_tuples.cpp]: We do this for ONNX export, so it is desired to flatten them in this case 😕

**Review by apaszke — APPROVED** [review]

**Review by jamesr66a — APPROVED** [review]
ONNX changes look functionally equivalent so LGTM
